source/equipment/equipment_api.cpp

```diff
@@ -0,0 +1,99 @@
+it under the terms of the GNU Affero General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Affero General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License
+along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+Stay tuned using
+twitter @navitia
+IRC #navitia on freenode
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+
+#include "equipment_api.h"
+#include "utils/paginate.h"
+
+#include "boost/range/algorithm.hpp"
+
+#include <map>
+#include <vector>
+#include <string>
+
+namespace navitia {
+namespace equipment {
+
+namespace {
+    const std::string build_ptref_line_filter(const std::string& line_uri, const std::string& filter) {
+        std::string line_filter = "line.uri=" + line_uri;
+        if(filter.size()) {
+            line_filter += " and (" + filter + ")";
+        }
+        return line_filter;
+    }
+
+    void fill_equipment_to_pb(const equipment::StopAreasPerLine& sas_per_line,
+                              PbCreator& pb_creator, int depth) {
+        for(const auto& line : sas_per_line) {
+            auto report = pb_creator.add_equipment_reports();
+            pb_creator.fill(line.first, report->mutable_line(), depth);
+            for(const auto& sa : line.second) {
+                auto sa_equip = report->add_stop_area_equipments();
+                pb_creator.fill(sa, sa_equip->mutable_stop_area(), depth);
+            }
+        }
+    }
+}
+
+StopAreasPerLine get_stop_areas_per_line(const type::Data& data, const std::string& filter,
+                                         const ForbiddenUris& forbidden_uris) {
+    StopAreasPerLine res;
+    const type::Indexes line_indices = ptref::make_query(type::Type_e::Line, filter, forbidden_uris, data);
+    const auto lines = data.get_data<type::Line>(line_indices);
+    for(const auto line : lines) {
+        const std::string line_filter = build_ptref_line_filter(line->uri, filter);
+        const type::Indexes sa_indexes = ptref::make_query(type::Type_e::StopArea, line_filter, forbidden_uris, data);
+        res.emplace_back(line, data.get_data<type::StopArea>(sa_indexes));
+    }
+
+    return res;
+}
+
+void equipment_reports(PbCreator& pb_creator, const std::string& filter,
+                       int count, int depth, int start_page,
+                       const ForbiddenUris& forbidden_uris) {
+    const type::Data& data = *pb_creator.data;
+    try {
+        const auto sas_per_line = get_stop_areas_per_line(data, filter, forbidden_uris);
+        const auto paginated_sas = paginate(sas_per_line, count, start_page);
```

Comment:
That's done. 
Also, a much better optimisation would be to pass an array of values in the ptref filter; like 
```sh
stop_point.has_code_type("TCL_escalier", "TCL_assenceur", "TCL_catapulte")
``` 
instead of 
```sh
stop_point.has_code_type("TCL_escalier") and stop_point.has_code_type("TCL_assenceur") and ....
```